db/event_manager_test.go

```diff
@@ -831,3 +831,126 @@ func assertChannelLengthWithTimeout(t *testing.T, c chan interface{}, expectedLe
 		}
 	}
 }
+
+// Simulate a valid event posting scenario of both online and offline database change event. Start event manager with
+// 10 processes and wait time of 100 seconds. Create 10 database state change online events and 10 offline events. Add
+// filter function to logic to post the events if state is offline; It should ignore the online events and post all the
+// online events; total of 10 online events should be posted.
+func TestWebhookHandleEventDBStateChange(t *testing.T) {
+	ts, wr := InitWebhookTest()
+	defer ts.Close()
+	url := ts.URL
+	wr.Clear()
+	em := NewEventManager()
+	em.Start(5, 100)
+
+	source := `function(doc) { if (doc.state == "offline") { return true; } else { return false; } }`
+	timeout := uint64(30)
+	wh, _ := NewWebhook(fmt.Sprintf("%s/echo", url), source, &timeout)
+	em.RegisterEventHandler(wh, DBStateChange)
+
+	warnCountPrev, err := strconv.Atoi(base.StatsResourceUtilization().Get(base.StatKeyWarnCount).String())
+	assert.NoError(t, err, "No error while getting the count of warnings")
+
+	for i := 0; i < 20; i += 2 {
+		err := em.RaiseDBStateChangeEvent("db", "online", "Index service is listening", "127.0.0.1:4985")
+		assert.NoError(t, err, "Database state change event should be raised successfully")
+		err = em.RaiseDBStateChangeEvent("db", "offline", "Index service heartbeat lost!", "127.0.0.1:4985")
+		assert.NoError(t, err, "Database state change event should be raised successfully")
+	}
+	func() {
+		timeout := time.After(1000 * time.Millisecond)
+		tick := time.Tick(100 * time.Millisecond)
+		for {
+			select {
+			case <-timeout:
+				log.Fatalf("Retry timedout while reading request count and payload!")
+				return
+			case <-tick:
+				if wr.GetCount() == 10 && len(wr.GetPayloads()) == 10 {
+					for i, s := range wr.GetPayloads() {
+						fmt.Printf("Posted payload-%v:%v\n", i, string(s))
+						assert.NotEmpty(t, string(s), "Request body should not be empty")
+						assert.Contains(t, string(s), `"state":"offline"`)
+						assert.NotContains(t, string(s), `"state":"online"`)
+					}
+					return
+				}
+			}
+		}
+	}()
+	warnCountCurr, err := strconv.Atoi(base.StatsResourceUtilization().Get(base.StatKeyWarnCount).String())
+	assert.NoError(t, err, "No error while getting the count of warnings")
+	warnCount := warnCountCurr - warnCountPrev
+	assert.Equal(t, 10, warnCount, "10 online events should be skipped and 10 offline events should be posted")
+}
+
+// Simulate the scenario for filter function error. Create 10 database status change (online) events with bad
```

Comment:
Comment doesn't match (we're no longer creating 10 events)

Reply:
This is fixed.